src/shinylive.ts

```diff
@@ -0,0 +1,512 @@
+    allFilesSorted.map(async (file: vscode.Uri): Promise<ShinyliveFile> => {
+      const type = isBinary(file.fsPath) ? "binary" : "text";
+      let name = pathRelativeToRootDir(file);
+
+      // Primary file needs to be `app.R` or `app.py` (or ui/server.R)
+      if (file.path === primaryFile) {
+        if (isPythonApp) {
+          name = "app.py";
+        } else if (isRApp && name.indexOf("app") >= 0) {
+          name = "app.R";
+        }
+      }
+
+      const contentRaw = await vscode.workspace.fs.readFile(file);
+      const content =
+        type === "binary"
+          ? Buffer.from(contentRaw).toString("base64")
+          : Buffer.from(contentRaw).toString();
+
+      switch (type) {
+        case "binary":
+          return { name, content, type };
+
+        default:
+          // Save some characters by relying on the implicit type "text" default
+          return { name, content };
+      }
+    })
+  );
+
+  await createAndOpenShinyliveApp(files, isPythonApp ? "py" : "r");
+}
+
+/**
+ * Read a directory recursively and return a list of all files as `vscode.Uri`
+ * objects. Directories whose names start with `_` or `.` are ignored.
+ *
+ * @async
+ * @param {vscode.Uri} uri A `vscode.Uri` of any type. If not a directory, the
+ * function returns an array with the single `vscode.Uri` object.
+ * @returns {Promise<vscode.Uri[] | void>}
+ */
+async function readDirectoryRecursively(
+  uri: vscode.Uri
+): Promise<vscode.Uri[] | void> {
+  const pathStat = await vscode.workspace.fs.stat(uri);
+  if (pathStat.type !== vscode.FileType.Directory) {
+    return [uri];
+  }
+
+  // Treat `_dir/` or `.dir/` as hidden directories and discard them
+  if (["_", "."].some((chr) => path.basename(uri.path).startsWith(chr))) {
+    return;
+  }
+
+  const files: vscode.Uri[] = [];
+  const filesInDir = await vscode.workspace.fs.readDirectory(uri);
+
+  for (const [name, type] of filesInDir) {
+    if (type === vscode.FileType.Directory) {
+      const subdir = vscode.Uri.file(`${uri.path}/${name}`);
+      const subdirFiles = await readDirectoryRecursively(subdir);
+      if (subdirFiles) {
+        files.push(...subdirFiles);
+      }
+    } else {
+      files.push(vscode.Uri.file(`${uri.path}/${name}`));
+    }
+  }
+
+  return files;
+}
+
+/**
+ * Consult the user's preferred open action, or ask them directly. The default
+ * preference is `"ask"`, but users can choose `"open"` or `"copy"` in their
+ * settings to avoid the prompt.
+ *
+ * @async
+ * @returns {Promise<UserOpenAction>} One of `"open"` or `"copy"`.
+ */
+async function askUserForOpenAction(): Promise<UserOpenAction> {
+  // first check if the user has set a default action
+  const prefAction =
+    vscode.workspace
+      .getConfiguration("shiny.shinylive")
+      .get<string>("openAction") || "ask";
+
+  if (["open", "copy"].includes(prefAction)) {
+    return prefAction as UserOpenAction;
+  }
+
+  const action = await vscode.window.showQuickPick(["open", "copy"], {
```

Comment:
For all these `showQuickPick` calls, it looks like if the answer is falsy, you're defaulting to a result. Doesn't a falsy answer mean that the user hit Escape or something? And therefore the whole operation should be cancelled?

Reply:
Yeah you're right. I learned about the canceling behavior after writing these two and forgot to go back and to update them.